### PR TITLE
Implement correct `:disabled` support for fieldsets and nested form controls

### PR DESCRIPTION
**Pull request by camchenry**
- Related to https://github.com/dperini/nwsapi/issues/97
- Related to https://github.com/jsdom/jsdom/issues/3232
- Related to https://github.com/jsdom/jsdom/issues/3603

With the current implementation of the `:disabled` selector, there are two main issues:

1. We are not considering nested fieldsets, only checking the first ancestor fieldset. The [HTML spec](https://html.spec.whatwg.org/#enabling-and-disabling-form-controls:-the-disabled-attribute) specifies that it should be any descendant:
    > A form control is disabled if any of the following are true: [...] the element is a descendant of a [fieldset](https://html.spec.whatwg.org/#the-fieldset-element) element whose [disabled](https://html.spec.whatwg.org/#attr-fieldset-disabled) attribute is specified, and is not a descendant of that [fieldset](https://html.spec.whatwg.org/#the-fieldset-element) element's first [legend](https://html.spec.whatwg.org/#the-legend-element) element child, if any.
2. We are not correctly considering fieldsets that do not have a `legend`. Currently, the check for if the legend element contains the element doesn't handle if the first legend element doesn't exist in the fieldset. Specifically, `(n=s.ancestor("fieldset",e))&&(n=s.first("legend",n))&&!n.contains(e))` returns false because `n=s.first("legend",n)` evaluates to `null` which is falsey, so the whole AND expression evaluates to false. Rather, if there is no legend element, we do not need to consider if the element is contained within the legend element and can solely focus on whether it is disabled or a descendant of a disabled fieldset.

Since the original `test/w3c/html/semantics/selectors/pseudo-classes/disabled.html` web-platform-test was added, it has been updated with additional test-cases. So I have added those here to ensure this works correctly.

### CODE REVIEW

File: src/nwsapi.js [hunk 1]
```diff
@@ -1123,10 +1123,31 @@
...
-                  // https://www.w3.org/TR/html5/forms.html#enabling-and-disabling-form-controls:-the-disabled-attribute
-                  source = 'if((("form" in e||/^optgroup$/i.test(e.localName))&&"disabled" in e&&' +
-                    '(e.disabled===true||(n=s.ancestor("fieldset",e))&&(n=s.first("legend",n))&&!n.contains(e))' +
-                    ')){' + source + '}';
+                  // https://html.spec.whatwg.org/#enabling-and-disabling-form-controls:-the-disabled-attribute
+                  source = 'if((("form" in e||/^optgroup$/i.test(e.localName))&&"disabled" in e)){' +
+                    // F is true if any of the fieldset elements in the ancestry chain has the disabled attribute specified
+                    // L is true if the first legend element of the fieldset contains the element
+                    'var x=0,N=[],F=false,L=false;' +
+                    'if(!(/^(optgroup|option)$/i.test(e.localName))){' +
```
Comment (camchenry) [inline src/nwsapi.js:1131]: This feels a bit messy, but `optgroup` has different definitions for `disabled`, as the HTML spec notes in the `disabled` attribute documentation:

> The [disabled](https://html.spec.whatwg.org/#attr-option-disabled) attribute for [option](https://html.spec.whatwg.org/#the-option-element) elements and the [disabled](https://html.spec.whatwg.org/#attr-optgroup-disabled) attribute for [optgroup](https://html.spec.whatwg.org/#the-optgroup-element) elements are defined separately.

We shouldn't take into account fieldset/legend for optgroup/option, so we skip over code related to that here.

**Comment by dperini** [comment]
@camchenry 
very well this is fantastic, this is  real help for me.
Thank you so much for the help catching the problem and code to fix it.
It will be merged immediately after a small review once tested against wpt.

**Comment by dperini** [comment]
@camchenry 
it is better if I apply your fix as is, without further delays since there have been enough testing.
Optimizations can be applied later if they still make this change to pass wpt requirements.

**Comment by dperini** [comment]
@camchenry & @kmcgrady
Closing this issue since it was completed and fixed by the merged pull request.
Thank you for the help and the contributed fix.

**Pull request merged**